staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -866,6 +872,44 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		}
 	}
 
+	if opts.requireMapListKeysMapSetValidation {
+		allErrs = append(allErrs, validateMapListKeysMapSet(schema, fldPath)...)
+	}
+
+	return allErrs
+}
+
+func validateMapListKeysMapSet(schema *apiextensions.JSONSchemaProps, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if schema.XListType != nil && (*schema.XListType == "map" || *schema.XListType == "set") && schema.Items != nil && schema.Items.Schema != nil {
+		is := schema.Items.Schema
+		if is.Nullable {
+			allErrs = append(allErrs, field.Forbidden(fldPath.Child("items").Child("nullable"), "cannot be nullable when x-kubernetes-list-type is "+*schema.XListType))
+		}
+
+		if *schema.XListType == "map" {
+			requiredMap := make(map[string]struct{}, len(is.Required))
```

Comment:
This part of the function could probably be implemented without allocation the map, especially since the XListMapKeys is really really short  (1 or 2 strings at most). I'm sure performance are not super important at this point, since it only happens on creation/update of CRDs which are not super frequent.

Reply:
I wouldn't care about this single allocation though. We do so many for every deep copy of the CRs anyway.